### PR TITLE
Allow dbloadrecordsloop to handle nesed macros

### PR DESCRIPTION
**Pull request by LowriJenkins**
dbloadrecordsloop changed to be able to handle macros of the form `\$(macro2\$(I))`. Also changed to more readable method of handling these macros.
https://github.com/ISISComputingGroup/IBEX/issues/7457
To test:
Change a call to dbloadrecordsloop  so that it uses a nested macro, e.g.
`EPICS\ioc\master\HUBER\iocBoot\iocHUBER-IOC-01\st-motor.cmd` line 36 change the call so that `AXIS=$(IOCNAME):\$(AXIS\$(I)=AXIS\$(I))`
Check that the IOC starts (if you use the HUBER_01 example you may need to start the emulator to check that it has started properly.

### CODE REVIEW

File: utilitiesApp/src/dbLoadRecordsFuncs.cpp [hunk 1]
```diff
@@ -66,26 +67,18 @@ static void loadMacEnviron(MAC_HANDLE* pmh)
...
-        new_macros.replace(start_pos, lf, loopSubTo);
-        start_pos += lt;
+    std::smatch bracketSearch, braceSearch;
+    // Check that we have done a search, and that the result of both is empty before finishing.
+    while((!bracketSearch.empty())||(!braceSearch.empty())||(!bracketSearch.ready())){
+        std::regex_search(new_macros, bracketSearch, std::regex(R"(\\\$\(([0-9a-zA-Z_\$\(\){}]+)\))"));
+        new_macros = std::regex_replace(new_macros, std::regex(R"(\\\$\(([0-9a-zA-Z_\$\(\){}]+)\))"), "$($1)");
+        // Have to look twice to guarantee that brackets/curly braces are properly matched.
+        std::regex_search(new_macros, braceSearch, std::regex(R"(\\\$\{([0-9a-zA-Z_\$\(\){}]+)\})"));
+        new_macros = std::regex_replace(new_macros, std::regex(R"(\\\$\{([0-9a-zA-Z_\$\(\){}]+)\})"), "$($1)");
```
Comment (FreddieAkeroyd) [inline utilitiesApp/src/dbLoadRecordsFuncs.cpp]: Possibly `${$1}` as replacement to preserve bracket types, though `(` and `{` are generally interchangable in EPICS

File: utilitiesApp/src/dbLoadRecordsFuncs.cpp [hunk 2]
```diff
@@ -66,26 +67,18 @@ static void loadMacEnviron(MAC_HANDLE* pmh)
...
-	start_pos = 0;
-    while( (start_pos = new_macros.find(loopSubFrom, start_pos)) != std::string::npos ) 
-    {
-        new_macros.replace(start_pos, lf, loopSubTo);
-        start_pos += lt;
+    std::smatch bracketSearch, braceSearch;
+    // Check that we have done a search, and that the result of both is empty before finishing.
+    while((!bracketSearch.empty())||(!braceSearch.empty())||(!bracketSearch.ready())){
+        std::regex_search(new_macros, bracketSearch, std::regex(R"(\\\$\(([0-9a-zA-Z_\$\(\){}]+)\))"));
+        new_macros = std::regex_replace(new_macros, std::regex(R"(\\\$\(([0-9a-zA-Z_\$\(\){}]+)\))"), "$($1)");
```
Comment (FreddieAkeroyd) [inline utilitiesApp/src/dbLoadRecordsFuncs.cpp]: The `std::regex` expression is repeated, could this be taken out into a constant? Also I believe you can skip the `\` escape within a `[` construct as it knows you just want normal characters i.e.    `[0-9a-zA-Z_$(){}]`

File: utilitiesApp/src/dbLoadRecordsFuncs.cpp [hunk 3]
```diff
@@ -66,26 +67,18 @@ static void loadMacEnviron(MAC_HANDLE* pmh)
 	}
 }
 
-/// look for e.g. \$(I) and replace with $(I) so we can substitute later with macEnvExpand()
+/// look for e.g. \$() and replace with $() so we can substitute later with macEnvExpand()
 static void subMacros(std::string& new_macros, const char* macros, const char* loopVar)
 {
```
Comment (FreddieAkeroyd) [inline utilitiesApp/src/dbLoadRecordsFuncs.cpp:72]: The `loopVar` argument could now be removed

**Comment by FreddieAkeroyd** [comment]
It would be nice to have some tests for this like in `https://github.com/ISISComputingGroup/EPICS-utilities/tree/master/utilitiesApp/tests` otherwise it is quite hard to test all the various edge cases. This may mean you need to split `subMacros` into a separate file so it can be included in the test suite.

**Review by pheest — APPROVED** [review]
